gax/src/main/java/com/google/api/gax/rpc/ReframingResponseObserver.java

```diff
@@ -0,0 +1,368 @@
+      deferError = false;
+      error = new CancellationException("User cancelled stream");
+    }
+
+    deliver();
+  }
+
+  /**
+   * Process a new response from inner/upstream callable. The message will be fed to the reframer
+   * and the output will be delivered to the downstream {@link ResponseObserver}.
+   *
+   * <p>If the delivery loop is stopped, this will restart it.
+   */
+  @Override
+  public void onResponse(InnerT response) {
+    synchronized (lock) {
+      awaitingInner = false;
+      reframer.push(response);
+    }
+    deliver();
+  }
+
+  /**
+   * Process inner/upstream callable's onError notification. This will be queued to be delivered
+   * after the delegate exhausts it's buffers.
+   *
+   * <p>If the delivery loop is stopped, this will restart it.
+   */
+  @Override
+  public void onError(Throwable t) {
+    synchronized (lock) {
+      if (error == null) {
+        error = t;
+      }
+      closeOnDone = true;
+    }
+    deliver();
+  }
+
+  /**
+   * Process inner/upstream callable's onComplete notification. This will be queued to be delivered
+   * after the delegate exhausts it's buffers.
+   *
+   * <p>If the delivery loop is stopped, this will restart it.
+   */
+  @Override
+  public void onComplete() {
+    synchronized (lock) {
+      closeOnDone = true;
+    }
+    deliver();
+  }
+
+  /**
+   * The message pump that interacts with the {@link this#outerResponseObserver}. All interactions
+   * with outer {@link ResponseObserver} will be serialized via this method.
+   */
+  /** Tries to kick off the delivery loop, wrapping it in error handling. */
+  private void deliver() {
+    // Ensure mutual exclusion via the inDelivery flag, if there is a currently active delivery
+    // then use the missed flag to schedule an extra delivery run.
+    synchronized (lock) {
+      if (closed || inDelivery) {
+        missed = true;
+        return;
+      }
+      inDelivery = true;
+    }
+
+    try {
+      unsafeDeliver();
+    } catch (Throwable t) {
+      outerResponseObserver.onError(t);
+    }
+  }
+
+  /** The internal states of the delivery loop */
+  private enum DeliveryAction {
+    /** There is supply & demand, so deliver the current message */
+    DELIVER,
+    /** There is demand but no supply, so request more from upstream */
+    REQUEST_MORE,
+    /** There is demand but no supply and with an outstanding request , so do nothing */
+    AWAIT_MORE_DATA,
+    /** Demand has been fully supplied */
+    FULFILLED,
+    /** The stream should be closed for various reasons */
+    CLOSE
+  }
+
+  /**
+   * Coordinates state transfer between inner/downstream callable and the outer/upstream. It
+   * orchestrates the flow of demand from downstream to upstream. The data flow from upstream
+   * through the delegate to downstream. It's back pressure aware and will only send as many
+   * messages that were requested. However it will send unsolicited onComplete & onError messages.
```

Comment:
s/that were/as were/, add a comma after "However"

Reply:
done